client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java

```diff
@@ -37,19 +38,13 @@
     private static final RequestFutureHolder INSTANCE = new RequestFutureHolder();
     private ConcurrentHashMap<String, RequestResponseFuture> requestFutureTable = new ConcurrentHashMap<String, RequestResponseFuture>();
     private final AtomicInteger producerNum = new AtomicInteger(0);
```

Comment:
there is no need use AtomicInteger.

I suggest change to HashSet<DefaultMQProducerImpl> and startScheduledTask/shutdown support re-entry. 

eg: called DefaultMQProducerImpl.shutdown got exception, and the use may call shutdown again.

Reply:
Don't understand what you mean

Reply:
like this
if(set.size > 0 && state != running) {
   startScheduledTask
}

if(set.size<=0 && state != shutdown){
  shutdown 
}